src/ring_theory/adjoin_root.lean

```diff
@@ -89,18 +89,48 @@ quotient.induction_on' x ih
 /-- Embedding of the original ring `R` into `adjoin_root f`. -/
 def of : R →+* adjoin_root f := (mk f).comp C
 
-instance [comm_semiring S] [algebra S R] : algebra S (adjoin_root f) :=
-ideal.quotient.algebra S
+instance [distrib_smul S R] [is_scalar_tower S R R] : has_smul S (adjoin_root f) :=
+submodule.quotient.has_smul' _
+
+instance [distrib_smul S R] [is_scalar_tower S R R] : distrib_smul S (adjoin_root f) :=
+submodule.quotient.distrib_smul' _
+
+@[simp]
+lemma smul_mk [distrib_smul S R] [is_scalar_tower S R R] (a : S) (x : R[X]) :
+  a • mk f x = mk f (a • x) := rfl
 
-instance [comm_semiring S] [comm_semiring K] [has_smul S K] [algebra S R] [algebra K R]
-  [is_scalar_tower S K R] :
-  is_scalar_tower S K (adjoin_root f) :=
+lemma smul_of [distrib_smul S R] [is_scalar_tower S R R] (a : S) (x : R) :
+  a • of f x = of f (a • x) :=
+by rw [of, ring_hom.comp_apply, ring_hom.comp_apply, smul_mk, smul_C]
+
+instance (R₁ R₂ : Type*) [has_smul R₁ R₂] [distrib_smul R₁ R] [distrib_smul R₂ R]
+  [is_scalar_tower R₁ R R] [is_scalar_tower R₂ R R] [is_scalar_tower R₁ R₂ R] (f : R[X]) :
+  is_scalar_tower R₁ R₂ (adjoin_root f) :=
 submodule.quotient.is_scalar_tower _ _
 
-instance [comm_semiring S] [comm_semiring K] [algebra S R] [algebra K R] [smul_comm_class S K R] :
-  smul_comm_class S K (adjoin_root f) :=
+instance (R₁ R₂ : Type*) [distrib_smul R₁ R] [distrib_smul R₂ R]
+  [is_scalar_tower R₁ R R] [is_scalar_tower R₂ R R] [smul_comm_class R₁ R₂ R] (f : R[X]) :
+  smul_comm_class R₁ R₂ (adjoin_root f) :=
 submodule.quotient.smul_comm_class _ _
 
+instance [monoid S] [distrib_mul_action S R] [is_scalar_tower S R R] (f : R[X]) :
+  distrib_mul_action S (adjoin_root f) :=
+submodule.quotient.distrib_mul_action' _
+
+instance [comm_semiring S] [algebra S R] : algebra S (adjoin_root f) :=
+ideal.quotient.algebra S
+
+-- TODO (lean4): make a smul lemma for `ideal.quotient.mk`.
+instance adjoin_root.is_scalar_tower_right [distrib_smul S R] [is_scalar_tower S R R] :
+  is_scalar_tower S (adjoin_root f) (adjoin_root f) :=
+⟨begin
+  rintro x ⟨y⟩ ⟨z⟩,
+  calc ideal.quotient.mk (span {f}) (x • y) * (ideal.quotient.mk (span {f})) z
+       = submodule.quotient.mk (x • y * z) : (_root_.map_mul (ideal.quotient.mk _) _ _).symm
+  ...  = x • (ideal.quotient.mk (span {f}) y * ideal.quotient.mk (span {f}) z) :
+  by rw [smul_mul_assoc, submodule.quotient.mk_smul, ideal.quotient.mk_eq_mk, _root_.map_mul]
+end⟩
```

Comment:
I need to wait for a cache, but I'm pretty sure this is just `ideal.quotient.is_scalar_tower_right` now

Reply:
Uh oh, this is causing trouble. I've split it to #18934 to see if it works in isolation.